thrust/thrust/system/tbb/detail/reduce_by_key.inl

```diff
@@ -63,18 +64,23 @@ inline L divide_ri(const L x, const R y)
 template <typename InputIterator, typename BinaryFunction, typename SFINAE = void>
 struct partial_sum_type
 {
-  using type = typename thrust::iterator_value<InputIterator>::type;
+  using value_t = thrust::iterator_value_t<InputIterator>;
+  // TODO(bgruber): unify with CUB's accumulator_t
+  using type = ::cuda::std::__decay_t<typename ::cuda::std::__invoke_of<BinaryFunction, value_t, value_t>::type>;
 };
 
+// gevtushenko preferred to detect a ::result_type on the function first
+_CCCL_SUPPRESS_DEPRECATED_PUSH
 template <typename InputIterator, typename BinaryFunction>
 struct partial_sum_type<InputIterator, BinaryFunction, ::cuda::std::void_t<typename BinaryFunction::result_type>>
 {
   using type = typename BinaryFunction::result_type;
 };
+_CCCL_SUPPRESS_DEPRECATED_POP
```

Comment:
important: I'm sorry for the confusion. In the offline discussion, I was under impression that checking `::result_type` of a binary operator was uniformly present across execution policies of reduce by key. But as you noted on the PR description, `::result_type` is only used in TBB backend. The idea of getting fundamentally different results when changing backends looks like a bug to me. 
```suggestion
```